python/src/pyarrow/adapters/pandas.cc

```diff
@@ -369,6 +392,10 @@ inline Status ArrowSerializer<NPY_OBJECT>::Convert(std::shared_ptr<Array>* out)
 
   // TODO: mask not supported here
   const PyObject** objects = reinterpret_cast<const PyObject**>(PyArray_DATA(arr_));
+  {
+    PyAcquireGIL lock;
+    PyDateTime_IMPORT;
```

Comment:
You're right, we should acquire the GIL below before using any of the Py* methods. I'll create a JIRA about going through this module completely and making sure we're releasing the GIL on the Cython side and re-acquiring it in the appropriate places here